viewer/packages/tools/src/Measurement/MeasurementTool.ts

```diff
@@ -0,0 +1,286 @@
+
+  private readonly _handleonPointerClick = this.onPointerClick.bind(this);
+  private readonly _handleonPointerMove = this.onPointerMove.bind(this);
+
+  private readonly _lineOptions: MeasurementLineOptions = {
+    lineWidth: 0.002,
+    color: 0x00ffff
+  };
+
+  constructor(viewer: Cognite3DViewer, options?: MeasurementOptions) {
+    super();
+    this._viewer = viewer;
+    this._options = options ?? {};
+    this._measurementLabel = new MeasurementLabels(this._viewer);
+    this._linePosition = new Float32Array(6);
+    this._pointSize = 0.02;
+    this._distanceValue = '';
+    this._distanceToCamera = 0;
+  }
+
+  /**
+   * Enter into point to point measurement mode.
+   */
+  enterMeasurementMode(): void {
+    this.setupEventHandling();
+  }
+
+  /**
+   * Exit measurement mode.
+   */
+  exitMeasurementMode(): void {
+    //remove measurement label, clear all mesh, geometry & event handling.
+    this._measurementLabel.clearLabels();
+    this.clearLineObjects();
+    this.removeEventHandling();
+  }
+
+  /**
+   * Sets Measurement line width and color with @options value.
+   * @param options MeasurementLineOptions to set line width and color.
+   */
+  setLineOptions(options: MeasurementLineOptions): void {
+    this._lineOptions.lineWidth = options?.lineWidth ?? this._lineOptions.lineWidth;
+    this._lineOptions.color = options?.color ?? this._lineOptions.color;
+    this._pointSize = options?.lineWidth * 10.0 || this._pointSize;
+  }
+
+  /**
+   * Dispose Measurement Tool.
+   */
+  dispose(): void {
+    this.exitMeasurementMode();
+    super.dispose();
+  }
+
+  /**
+   * Set input handling.
+   */
+  private setupEventHandling() {
+    this._viewer.on('click', this._handleonPointerClick);
+  }
+
+  /**
+   * Remove input handling.
+   */
+  private removeEventHandling() {
+    this._viewer.off('click', this._handleonPointerClick);
+  }
+
+  /**
+   * Clear all line objects mesh, geometry & material.
+   */
+  private clearLineObjects() {
+    if (this._lineMesh) {
+      this._lineMesh.geometry.dispose();
+      this._lineMesh = null;
+    }
+    if (this._lineGeometry) {
+      this._lineGeometry.dispose();
+      this._lineGeometry = null;
+    }
+    if (this._lineMaterial) {
+      this._lineMaterial.dispose();
+      this._lineMaterial = null;
+    }
+  }
+
+  private async onPointerClick(event: MouseEvent) {
+    const { offsetX, offsetY } = event;
+
+    const intersection = await this._viewer.getIntersectionFromPixel(offsetX, offsetY);
+
+    if (intersection) {
+      this.addSphere(intersection.point);
+
+      if (!this._lineMesh) {
+        this._viewer.domElement.addEventListener('mousemove', this._handleonPointerMove);
+        this.startMeasurement(intersection);
+      } else {
+        this.endMeasurement(intersection.point);
+        this._viewer.domElement.removeEventListener('mousemove', this._handleonPointerMove);
+      }
+      this._viewer.requestRedraw();
+    }
```

Comment:
NIT: Invert this if statement, as in if !intersection then return. Will reduce indentation / branching